app/src/main/java/de/markusfisch/android/shadereditor/hardware/MicInputListener.java

```diff
@@ -0,0 +1,132 @@
+import android.media.MediaRecorder;
+import android.os.Build;
+import android.os.Handler;
+import android.os.Looper;
+import android.util.Log;
+
+import androidx.core.content.ContextCompat;
+
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+
+public class MicInputListener extends AbstractListener {
+	private static final String TAG = "MicInputListener";
+	private static final int SAMPLE_RATE = 44100; // Hz
+	private static final int CHANNEL_CONFIG = AudioFormat.CHANNEL_IN_MONO;
+	private static final int AUDIO_FORMAT = AudioFormat.ENCODING_PCM_16BIT;
+	private static final int BUFFER_SIZE = AudioRecord.getMinBufferSize(
+			SAMPLE_RATE,
+			CHANNEL_CONFIG,
+			AUDIO_FORMAT);
+
+	private AudioRecord audioRecord;
+	private Thread recordingThread;
+	private volatile boolean isRecording = false;
+	private float currentAmplitude = 0f;
+
+	private final Context context;
+
+	public MicInputListener(Context context) {
+		super(context);
+		this.context = context;
+	}
+
+	@Override
+	public void unregister() {
+		if (!isRecording) {
+			return;
+		}
+
+		isRecording = false;
+		if (recordingThread != null) {
+			try {
+				recordingThread.join();
+			} catch (InterruptedException e) {
+				Log.e(TAG, "Error joining recording thread", e);
+			}
+			recordingThread = null;
+		}
+
+		if (audioRecord != null) {
+			audioRecord.stop();
+			audioRecord.release();
+			audioRecord = null;
+		}
+		currentAmplitude = 0f;
+	}
+
+	public boolean register() {
+
+
+		if (isRecording) {
+			return true;
+		}
+
+		if (ContextCompat.checkSelfPermission(this.context, Manifest.permission.RECORD_AUDIO)
+				!= PackageManager.PERMISSION_GRANTED) {
+			Log.e(TAG, "RECORD_AUDIO permission not granted.");
+			return false;
+		}
+
+		if (BUFFER_SIZE == AudioRecord.ERROR_BAD_VALUE || BUFFER_SIZE == AudioRecord.ERROR) {
+			Log.e(TAG, "AudioRecord.getMinBufferSize returned an error.");
+			return false;
+		}
+
+		audioRecord = new AudioRecord(
+				MediaRecorder.AudioSource.MIC,
+				SAMPLE_RATE,
+				CHANNEL_CONFIG,
+				AUDIO_FORMAT,
+				BUFFER_SIZE);
+
+		if (audioRecord.getState() != AudioRecord.STATE_INITIALIZED) {
+			Log.e(TAG, "AudioRecord initialization failed.");
+			audioRecord.release();
+			audioRecord = null;
+			return false;
+		}
+
+		isRecording = true;
+		recordingThread = new Thread(new Runnable() {
+			@Override
+			public void run() {
+				processAudioData();
+			}
+		});
+		recordingThread.start();
+
+		return true;
+	}
+
+	private void processAudioData() {
+		short[] buffer = new short[BUFFER_SIZE / 2]; // 16-bit PCM, so 2 bytes per sample
+		audioRecord.startRecording();
+
+		while (isRecording) {
+			int read = audioRecord.read(buffer, 0, buffer.length);
+			if (read > 0) {
+				long sum = 0;
+				for (int i = 0; i < read; i++) {
+					sum += Math.abs(buffer[i]);
+				}
+
+				currentAmplitude = (float) sum / read;
+
+				// Normalize amplitude to a 0-1 range, assuming max 16-bit value is 32767
+				currentAmplitude /= 32767.0f;
```

Comment:
This is run in a thread, so there is a very low chance, that the code reading `currentAmplitude` reads the value before the division. Instead, maybe use a local variable first:

```groovy
float newAmplitude = (float) sum / read;
newAmplitude /= (float) Short.MAX_VALUE;

currentAmplitude = newAmplitude;
```